sor/src/main/java/com/bazaarvoice/emodb/sor/db/astyanax/AstyanaxDataReaderDAO.java

```diff
@@ -390,24 +391,59 @@ public Record read(Key key, ReadConsistency consistency) {
 
-            splits.addAll(getCfSplits(keyspace, cf, start, end, desiredRecordsPerSplit, keyRange.getStart()));
+            splits.addAll(getCfSplits(keyspace, cf, start, end, desiredRecordsPerSplit, allTokenRanges));
         }
         return splits;
     }
 
     private List<CfSplit> getCfSplits(Keyspace keyspace, ColumnFamily<ByteBuffer, UUID> cf, String start,
-                                    String end, int desiredRecordsPerSplit, ByteBuffer startKey) {
-        try {
-            return keyspace.describeSplitsEx(cf.getName(), start, end, desiredRecordsPerSplit, startKey);
-        } catch (ConnectionException e) {
-            throw Throwables.propagate(e);
+                                    String end, int desiredRecordsPerSplit, Iterable<TokenRange> allTokenRanges) {
+        // There is a hole in the describeSplitsEx() call where if the call is routed to a Cassandra node which does
+        // have a replica of the requested token range then it will return a single split equivalent to the requested
+        // range.  To accommodate this each query is routed to a host that is verified to have a replica of the range.
+
+        ScanRange splitRange = ScanRange.create(parseTokenString(start), parseTokenString(end));
+        List<CfSplit> cfSplits = Lists.newArrayList();
+
+        // Iterate over the entire ring to find the token ranges which overlap with the provided range
+        for (TokenRange hostTokenRange : allTokenRanges) {
+            ScanRange hostSplitRange = ScanRange.create(
+                    parseTokenString(hostTokenRange.getStartToken()),
+                    parseTokenString(hostTokenRange.getEndToken()));
+
+            // Use the intersection to determine if there is overlap
+            for (ScanRange intersection : splitRange.intersection(hostSplitRange)) {
+                // Try once on each host until splits are returned
+
+                List<CfSplit> intersectionSplits = null;
+                for (Iterator<String> hosts = hostTokenRange.getEndpoints().iterator(); hosts.hasNext() && intersectionSplits == null; ) {
+                    String host = hosts.next();
+                    try {
+                        intersectionSplits = KeyspaceUtil.pin(keyspace).toHost(host)
+                                .describeSplitsEx(cf.getName(), toTokenString(intersection.getFrom()),
+                                        toTokenString(intersection.getTo()),
+                                        desiredRecordsPerSplit, intersection.getFrom());
+                    } catch (ConnectionException e) {
+                        // If there is another host to try then do so, otherwise raise the exception
+                        if (!hosts.hasNext()) {
+                            throw Throwables.propagate(e);
+                        }
+                    }
+
+                    assert intersectionSplits != null : "Exception would have been thrown if no splits had returned successfully";
```

Comment:
I think this assert is at a wrong place (may be be placed out of the loop after checking all hosts). This will raise an assertion error if the 1st host (out of the 3 hosts having the token range) had a connectionException. In this case, we would any way run on the next host.